jetty-core/jetty-client/src/test/java/org/eclipse/jetty/client/HttpClientResponseListenersTest.java

```diff
@@ -0,0 +1,201 @@
+            {
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public boolean onHeader(Response response, HttpField field)
+            {
+                // Number of header may vary, so don't count
+                return true;
+            }
+
+            @Override
+            public void onHeaders(Response response)
+            {
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onContent(Response response, ByteBuffer content)
+            {
+                // Should not be invoked
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onContent(Response response, Content.Chunk chunk, Runnable demander)
+            {
+                // Should not be invoked
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onSuccess(Response response)
+            {
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onFailure(Response response, Throwable failure)
+            {
+                // Should not be invoked
+                counter.incrementAndGet();
+            }
+
+            @Override
+            public void onComplete(Result result)
+            {
+                assertEquals(200, result.getResponse().getStatus());
+                counter.incrementAndGet();
+                latch.countDown();
+            }
+        };
+        client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            .onResponseBegin(listener)
+            .onResponseHeader(listener)
+            .onResponseHeaders(listener)
+            .onResponseContent(listener)
+            .onResponseContentAsync(listener)
+            .onResponseSuccess(listener)
+            .onResponseFailure(listener)
+            .onComplete(listener)
+            .send(listener);
+
+        assertTrue(latch.await(5, TimeUnit.SECONDS));
+        int expectedEventsTriggeredByResponseListeners = 4;
+        int expectedEventsTriggeredBySendListener = 4;
+        int expected = expectedEventsTriggeredByResponseListeners + expectedEventsTriggeredBySendListener;
+        assertEquals(expected, counter.get());
+    }
+
+    @ParameterizedTest
+    @ArgumentsSource(ScenarioProvider.class)
+    public void testResponseListenerAddedInResponseListenerEvent(Scenario scenario) throws Exception
+    {
+        start(scenario, new EmptyServerHandler());
+
+        CountDownLatch beginLatch = new CountDownLatch(1);
+        CountDownLatch headersLatch = new CountDownLatch(1);
+        ContentResponse response = client.newRequest("localhost", connector.getLocalPort())
+            .scheme(scenario.getScheme())
+            // Listeners for future events will be notified.
+            .onResponseBegin(r1 -> r1.getRequest().onResponseHeaders(r2 -> beginLatch.countDown()))
```

Comment:
Oh this cooks my noodle!    Let me see if I understand this correctly....
The `onResponseBegin` listener will be called when the first response in the conversation is called.  The `r1.getRequest()` gets the request for this response, which is the original/first response.   Thus the `onResponseHeaders` is adding another listener to the same request?

Would it be correct to say:
```java
assert r1.getRequest() == request_returned_from_newRequest;
assert r1 == r2;
assert r2.getRequest() == request_returned_from_newRequest;
```